woocommerce/woocommerce-gutenberg-products-block
# Product button: sync temporary number of items on instantiation

## What

I've fixed a sync problem between the number of items in the cart (`state.woocommerce.cart`) and the number of items that we show in the Product Button (`context.woocommerce.temporaryNumberOfItems`).

## Why

We use two values to be able to trigger an animation when those values are different (for example, when updated from the MiniCart).  When using client-side navigation, we want to show the number in `state.woocommerce.cart`, not in `context.woocommerce.temporaryNumberOfItems` because the cart is the source of truth. Also, we don't want to trigger an animation, we want to show it right away.

## How

I've used a `useLayoutEffect` to sync the `temporaryNumberOfItems` with the cart items when the element is instantiated and the cart (`state.woocommerce.cart`) has already loaded. I don't love it because it means manually syncing both states, which is something we should try to avoid whenever possible.

Anyway, I'll think more about this pattern, but this should fix it for now.

It also adds `key`'s to the Product Collection array.














### Screenshots



<div>
    <a href="https://www.loom.com/share/c43275ccb80d4286b96da2c3b75e9280">
      <img style="max-width:300px;" src="https://cdn.loom.com/sessions/thumbnails/c43275ccb80d4286b96da2c3b75e9280-with-play.gif">
    </a>
  </div>

https://www.loom.com/share/c43275ccb80d4286b96da2c3b75e9280

### Testing

1. Add a Product Collection block
2. Add items to the cart
3. Paginate to a different page
4. Add more items to the cart
5. Check that when paginating back and forth, it shows the correct number of items
6. Open the MiniCart block
7. Change the number of items of products that are in the same page
8. Check that the animation is triggered
9. Change the number of items of products that are not in the same page
10. Navigate to the page that has those products and check that the number of items is correct
11. Use a cache plugin (or simulate a cached page)
12. Check everything again

### WooCommerce Visibility




## Checklist

Required:

Conditional:





## Changelog


> Fixed a sync problem between the number of items in the car and the number of items that we show in the Product Button
